### ISSUE TITLE
Additional Protocol Support

### ISSUE THREAD

**Issue by jkowall** [w3c/trace-context-protocols-registry#4]
Today's OSS tracing systems are focused on http, the typical enterprise has a lot of other protocols which are critical to the application. Some examples include:

**1. JMS**
	Data can be embedded in message properties without harm, and this allows correlation and identification of the presence of a messaging system. The data can be tagged going in and measured for message put. Similarly, the receiving end can measure the time in the queue and the time it takes to process said message.

**2. Native MQ**
	This is more difficult to handle since native MQ protocols do not have specific areas where message data can be inserted without the fear of breaking the downstream legacy code. There are several tricks here which can help avoid the problem, and there are published recommendations from IBM, Tibco, and others. The problem is that the downstream applications typically ignore the recommendations. 

**3. AMQP**
	Similar to JMS, supporting AMQP is possible due to having extensible message formats. This still must be done by the library or programmer to inject the right data into the message.  

**4. DB Calls (JDBC / ADO.NET / other DB access types)**
	Although this is not something I've seen tracing tools do today, it could be of use to have similar data going to the database. This is something that Oracle has done internally, for example, the database can automatically record a transaction id with the record for trackability (https://docs.oracle.com/cd/B19306_01/java.102/b14355/endtoend.htm#BEICGHJH). It would depend on the driver and database but could be of use. 

**5. RMI or other RPC**
	In Java, these calls would typically be via CORBA or via raw calls to a downstream system. For things like CORBA, there are fields which are available to do this type of thing. Similarly, on Hadoop RPC (https://wiki.apache.org/hadoop/HadoopRpc) this would be possible as well. Some of these would require adding data to a message on one end, and possibly removing it on the remote end to avoid breakage.

**6. Protobuf and other binary protocols**
	These are potentially interesting technologies we could embed into, but once again it depends if we can remove the data before it's processed downstream, or find a way to embed data into the payload without causing issues. This is currently a challenge for APM tools.

**Comment by kag0** [comment]
Over [on the JSON-RPC mailing list](https://groups.google.com/g/json-rpc/c/l7m5SIvwnFs) we've been talking about creating a trace context for JSON-RPC. 
Is there a way we can get a repo under the w3c org (w3c/trace-context-json-rpc) to start drafting?
Please let me know if there's another, better place to discuss.

**Comment by SergeyKanzhelev** [comment]
@kag0 if you want to host it someplace else - you can follow this example: https://github.com/w3c/trace-context-protocols-registry/pull/11

Having spec under w3c org may create unnecessary overhead for you, but if needed I believe it can be done. @caribouW3 any objections?

**Comment by kag0** [comment]
@SergeyKanzhelev we'd prefer the formality of having it as a w3c spec, but open to alternate directions as we progress.

**Comment by SergeyKanzhelev** [comment]
@kag0 added to the agenda of the next meeting: https://docs.google.com/document/d/12AAIbnkt_AH1YjZwUwxirX_7WA7aWOSXpDweCgfqIiQ/edit#heading=h.g9nd8akmp2p2 Next meeting on calendar is Dec 1st.

My understanding is that nothing should prevent us from having this repository. I believe contributors/editors for the spec must be in w3c org which may be a blocker for some people to participate (just something to consider).

**Comment by kag0** [comment]
Yes that might be a bit of a blocker. @AArnott and I are the two people working on it, and neither of us are in w3c. 
So one or both of us would need to join w3c if we wanted to author content for the spec? ie. we can't open PRs and have someone inside the org merge them?

**Comment by AArnott** [comment]
How do we join the w3c?

**Comment by caribouW3** [comment]
@AArnott your company is already a W3C Member, you only need the Advisory Committee representative of the company to join the group and nominate you as a participant.
After that you can contribute to the group's work (this is for securing the licensing/patent commitments)

It is also possible to use the W3C submission process https://www.w3.org/2020/Process-20200915/#Submission

**Comment by AArnott** [comment]
I haven't had any luck finding who at Microsoft is an Advisory Committee rep. You wouldn't happen to have a list of those, would you?

**Comment by caribouW3** [comment]
First you need to create a W3C account https://www.w3.org/accounts/request, affiliated to Microsoft. Then ask @travisleithead (AC rep)

**Comment by flx5** [comment]
Would it make sense to add SOAP to the List? Or is one supposed to instrument the underlying protocol (http for example)?

**Comment by KalleOlaviNiemitalo** [comment]
For Apache Kafka, there is existing practice that `traceparent` and others are added as headers in the Kafka message, and their string values are encoded in UTF-8.  The [binary format](https://w3c.github.io/trace-context-binary/) is not used, even though Kafka header values are byte arrays and it would save space.  <https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/f11d176c15f6c8c712e1b611cb0ba6a9113581dd/instrumentation/kafka/kafka-clients/kafka-clients-common-0.11/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/common/v0_11/internal/KafkaHeadersSetter.java>

For SOAP, there is code that adds `traceparent` and others as HTTP request headers and also as SOAP headers.  The XML namespace URI of these SOAP headers is `https://www.w3.org/TR/trace-context/`.  <https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/fdedd5baf6ed03624fee7d43146e754e82733bd9/src/OpenTelemetry.Instrumentation.Wcf/Implementation/TelemetryPropagationWriter.cs>; and according to <https://github.com/CoreWCF/CoreWCF/issues/790#issuecomment-2215032863>, these SOAP headers are planned to be implemented in CoreWCF as well.

For correlated tracing of SOAP (including non-HTTP) clients and servers, there is also _[MS-NETTR]: .NET Tracing Protocol_ published by Microsoft, but that is based on GUIDs and cannot carry the entire W3C `traceparent`, much less `tracestate`.
